README.md
# Lifelong Learning Explorer (L2Explorer)
![logo](docs/apl_small_vertical_blue.png)

## Table of Contents

- [Setting up l2explorer_env](#introduction)
- [Trying out l2explorer_env](#metrics)
- [Testing Curricula for Continual Reinforcement Learning](#evaluation)
- [Benchmarking](#getting-started)
- [Procedurally generated worlds](#procedurally-generated-worlds)
- [Observation space](#observation-space)
- [Action Space](#action-space)
- [Observation Space](#observation-space)
- [Reward](#reward)
- [Episode Termination](#episode-termination)
- [Random Seed](#random-seed)
- [Changelog](#changelog)
- [License](#license)
- [Acknowledgements](#acknowledgements)

v1.0 Documentation
![diagram](l2explorer_environment.jpg)

## Setting up 'l2explorer_env'

Lifelong Learning Explorer (L2Explorer) is a 3D Unity environment that provides first-person-view observations and challenges for a lifelong learning agent, particularly tailored to test continual reinforcement learning. For more details (including how to download and install the Unity environment), see the [docs/Getting_Started_with_L2Explorer.md](docs/Getting_Started_with_L2Explorer.md) file. 

This repository (`l2explorer_env`) is a Python interface that provides a OpenAI Gym-like environment to L2Explorer. It is built on top of the Unity ML-Agents framework (https://unity.com/products/machine-learning-agents).

## Trying out `l2explorer_env`

To test out the `l2explorer_env` environment, run the `examples/random_agent.py` script. Two example jsons are provided to get started quickly.  Note that you will need to be in the `l2explorer_env/examples/` folder to run the following command or you may alternatively specificy the full path name for the .py and .json scripts.

```bash
python random_agent.py -reps 1 -maxsteps 200 -jsonfile map0.json
```

## Testing Curricula for Continual Reinforcement Learning

The key contribution of L2Explorer is a set of testing curricula for testing lifelong learning agents and characterizing sensitivities to different changes.

These curricula test reinforcement learning agents against continual and discrete shifts in the reward, state, state transition, observation, and action spaces. 

## Benchmarking

Results are logged with l2logger, and metrics calculated with l2metrics.

## Procedurally generated worlds

![Sample Map](examples/procedural_generation/example2_sample_map.png)

L2Explorer enables a high degree of procedural generation of worlds for the agent, including:

- types and appearances of a variety of objects
- locations and probabilistic relationships between objects
- how agents may interact with objects (bump with positive or negative reward, pick them up, etc.)

For more information, see:

- [docs/Outline.md](docs/Outline.md) for details of the procedural specification format
- `examples/random_agent.py` for an example of how the procedural generation may be sent to the L2Explorer environment (it loads in the specification as a JSON file)
- `examples/linear_agent.py` and `examples/waypoint_agent.py` show additional examples of agents for controlling motion
- `examples/procedural_generation/` for how the JSON file might be generated
- `examples/spawning_agent.py` shows how the sidechannels can be used to dynamically spawn objects in the environment
- `examples/l2ex_guiagent.py` for a sample "GUI agent" that allows a human to interact with the L2Explorer world
- examples/logging provides examples for incorporating l2logger into the code

## Observation space

More details on the action observation space, and reward options cna be found in the docs folder. The options enabled with the default maps and tasks are listed below. An agent for these configurations can be found in `examples/random_agent.py`.

## Action Space

The default motion model is ByVelocity. Further models are planned but are still being tested
Action space is a [3,1] vector.

- Element 0 is the linear velocity command, specified between -max_linear_velocity and +max_linear_velocity parameter
- Element 1 is the angular velocity command, specified between -max_angular_velocity and +max_angular_velocity parameter
- Element 2 is the pickup action, which activates a pick-up action if the value is above 0.5. Below 0.5 is a no-op.

## Observation Space

The state, by default, returns a visual and vector observation

- visual_obs = state["visual"] # list of visual observation
- state_vector = state["state"] # length 4 array with x,y,theta,velocity of agent in world frame

The visual observation contains 3 observations in a list- the first is a depth image, the second is a RGB image, third is a semantic segmentation.
All are shape [84,84,3] by default. Currently, images need to be resized in python code, and high-resolution observation cannot be produced by Unity.

## Reward

The object reward is set in the json for the environment, specified by setting the agent interaction parameters as specified in docs/outline.md. Supported interactions include collision, interaction, and in_range. Objects can also be destroyed with the same conditions.

## Episode Termination

By default, the episode terminates when "max_steps" is reached.

## Random Seed

You can control the seed by calling L2ExplorerTask.seed with the desired value. Otherwise, the default environment seed is 1234.

## Changelog

See [CHANGELOG.md](./CHANGELOG.md) for a list of notable changes to the project.

## License

See [LICENSE.md](LICENSE.md) for license information.

## Acknowledgements

Primary development of Lifelong Learning Explorer (L2Explorer) was funded by the DARPA Lifelong Learning Machines (L2M) Program.

© 2022 The Johns Hopkins University Applied Physics Laboratory LLC

## Disclaimer 

The views, opinions, and/or findings expressed are those of the author(s) and should not be interpreted as representing the official views or policies of the Department of Defense or the U.S. Government.

CHANGELOG.md
# l2explorer_env Changelog

All notable changes to this repository are documented here. We are using [Semantic Versioning for Documents](https://semverdoc.org/), in which a version number has the format `major.minor.patch`.

## 1.0.0

Updates to baseline agents and curricula for release

## 9.1.1

Updates to curriculum execution, inclusion of baselines. 

## 0.9.1 - 2021-7-13

Minor updates on documentation, unit8 support for images, bug fixes. Compatible with l2explorer049.

## 0.9.0 - 2021-1-20

Minor updates on documentation, spawning agent. Compatible with l2explorer047.

## 0.8.0 - 2020-12-20

Changes for split assets with l2explorer 047.

## 0.7.0 - 2020-11-09

Major changes for compatibility with L2explorer 045. This includes modifications to the side channels to support object creation, support for new modes of object interaction, and additional examples. A logging example is also provided with this version.

## 0.6.0 - 2020-08-08

Major changes to incorporate latest version of MLAgents (and use of side channel implementation)
Bug fixes and code refactoring

## 0.5.2 - 2020-03-06

- Updated documentation (docs/Outline.md) with defaults for spawnable objects
- Bug fixes

## 0.5.1 - 2020-01-25

- Updated examples and descriptions in docs/Outline.md
- Minor fixes to JSON templates for procedural generation

## 0.5.0 - 2019-12-05

Major revamp of design

- Support for procedurally generated environments in L2Explorer application
  - The `reset()` method takes a Python dict with a rich set of capabilities ([more info](docs/Outline.md))
- Procedural generation of stochastically co-occurring objects (`l2explorer/l2procgen/`)
- GUI agent that allows a human to play the part of the AI agent (`examples/l2ex_guiagent`)
- Tested + supported on Linux

## 0.1.0 - 2019-09-01

- Initial submission
- Support for parameterized L2Explorer tasks
- Singleton implementation, so that the Unity executable isn't reloaded for each task
- Tested on Windows and Linux


examples/spawning_agent.py
"""
Copyright © 2021 The Johns Hopkins University Applied Physics Laboratory LLC
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the “Software”), to
deal in the Software without restriction, including without limitation the
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
sell copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
"""

import argparse
import json
import math
import random
import sys

import l2explorer
import numpy as np
from l2explorer.l2explorer_env import L2ExplorerTask

"""
Sample code illustrating how to start up and interact with 
an L2Explorer instance without using the learnkit (syllabus) 
interface.
This example uses the sidechannels to spawn a new geometric object 
python spawning_agent.py -reps 3 -maxsteps 200 -jsonfile map_findobjects_0.json
"""
spawn_json = {
    "model": "1",
    "class": "geometric",
    "coordinates": [12.4, 10],
    "color": [1.0, 0.0, 0.0],
    "motion_model": "stationary",
    "destroy_stimulus": [
        "AgentCollide"
    ],
    "reward_once_stimulus": [
        "AgentCollide"
    ],
    "reward": 100.0
}

class SpawnAgent():
    def __init__(self):
        self._random = random.Random()

    def process_step(self, state):
        # action space: continuous values 
        # Linear movement: [-max_velocity,max_velocity]
        # Angular movement: [-max_angular_velocity,max_angular_velocity]
        # Pick up action: [-1,1] (if less than or equal to 0, no action is taken. 
        #    If greater than 0 the agent activates pick up action). Currently not used in examples given.

        #Default max values are linear speed 10, angular speed 90
        action = np.random.random_sample((3, 1))
        action[0] = 3.5  # action[0]*10
        action[1] = 0.0  # action[1]*90
        action[2] = 1.0
        return action


def play_spawn_agent(game, params, max_steps=500):
    """ Play a game until completion with random agent"""
    agent = SpawnAgent()
    done = False
    num_steps = 0
    state = game.reset(params)
    while not done and num_steps < max_steps:
        action = agent.process_step(state)
        state, reward, done, info = game.step(action)
        # Each Visual observation is default size 84,84,3
        # There are three visual observations by default
        # 0: Depth Map
        # 1: RGB View
        # 2: Semantic Segmentation
        visual_obs = state["visual"]  # list of visual observation
        state_vector = state["state"]  # length 4 array with x,y,theta of agent in world frame
        if reward > 0:
            print(reward)
            #spawn a new object on reward 10 units in front of agent
            #Theta = 0 for bot is along y axis. heading is in degrees
            coord_x = state_vector[0] + np.sin(2*math.pi*state_vector[2]/360.0) * 6
            coord_y = state_vector[1] + np.cos(2*math.pi*state_vector[2]/360.0) * 6
            spawn_json["coordinates"] = [coord_x, coord_y]
            print(spawn_json)
            print(state_vector)
            game.spawn_object(spawn_json)
        num_steps += 1


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Dummy agent for playing L2Explorer games")
    parser.add_argument('-reps', type=int, default=3, help="Number of times to play the game (def=3)")
    parser.add_argument('-maxsteps', type=int, default=200, help="Max number of time steps (def=200). Set to 0 to disable.")
    parser.add_argument('-jsonfile', type=str, help="File to environment JSON.")
    args = parser.parse_args()

    #load param
    with open(args.jsonfile) as json_file:
        parsed_json = json.load(json_file)

    max_steps = sys.maxsize if args.maxsteps == 0 else args.maxsteps

    game = L2ExplorerTask()
    try:
        for _ in range(args.reps):
            play_spawn_agent(game, params=parsed_json, max_steps=max_steps)
    finally:
        game.close_env()
